src/core/shape_inference/include/shape_infer_transformations.hpp

```diff
@@ -0,0 +1,59 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "compare.hpp"
+#include "openvino/core/except.hpp"
+
+namespace ov {
+namespace sh_infer {
+namespace tr {
+
+/**
+ * \brief Trnsform tensor data by cast them to type T
+ *
+ * \tparam T Type of returned value.
+ */
+template <class T>
+struct Cast {
+    constexpr Cast() = default;
+
+    template <class U>
+    constexpr T operator()(const U u) const {
+        return static_cast<T>(u);
+    }
+};
+
+/**
+ * \brief Check if input data is in [T::min(), T::max()] and then cast it to T.
+ *
+ * \tparam T Type of returned value and used to specified min, max of valid value range.
+ *
+ * \throws ov::AssertFailure if input value not in type range.
+ */
+template <class T>
+struct InTypeRange {
+    const std::pair<T, T> m_range{};
+    bool in_range{};
+
+    constexpr InTypeRange() : m_range{std::numeric_limits<T>::min(), std::numeric_limits<T>::max()} {};
+
+    template <class U>
+    T operator()(const U u) const {
+        OPENVINO_ASSERT(cmp::le(m_range.first, u) && cmp::le(u, m_range.second),
+                        "Value ",
+                        u,
+                        " not in range [",
+                        m_range.first,
+                        ":",
+                        m_range.second,
+                        "]");
+        return Cast<T>()(u);
```

Comment:
Could the `static_cast<T>` be used here instead? I don't see the benefit of creating an empty temporary object without any state.

Reply:
Done